core/kernel/notif.c

```diff
@@ -57,31 +80,42 @@ void notif_unregister_driver(struct notif_driver *ndrv)
 	cpu_spin_unlock_xrestore(&notif_lock, old_itr_status);
 }
 
-void notif_deliver_atomic_event(enum notif_event ev)
+void notif_deliver_atomic_event(enum notif_event ev, uint16_t guest_id)
 {
-	uint32_t old_itr_status = 0;
+	struct guest_partition *prtn = virt_get_guest(guest_id);
+	struct notif_data *ndata = get_notif_data(prtn);
 	struct notif_driver *nd = NULL;
-
-	assert(ev == NOTIF_EVENT_STARTED);
+	uint32_t old_itr_status = 0;
 
 	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
 
-	if (notif_started) {
-		DMSG("Already started");
-		goto out;
+	switch (ev) {
+	case NOTIF_EVENT_STARTED:
+		if (ndata->notif_started) {
+			DMSG("Already started");
+			return;
```

Comment:
`virt_put_guest()` to balance line 85?

Reply:
Indeed, I'll add a goto out, we must unlock the spinlock too.